dokka-integration-tests/gradle/src/testExampleProjects/kotlin/ExampleProjectsTest.kt

```diff
@@ -0,0 +1,344 @@
+        LibraryPublishing,
+        Multimodule,
+        VersioningMultimodule,
+        ;
+
+        companion object {
+            fun of(dir: Path): ExampleProject {
+                return when (dir.name) {
+                    "basic-gradle-example" -> BasicGradle
+                    "javadoc-example" -> Javadoc
+                    "composite-build-example" -> CompositeBuild
+                    "custom-styling-example" -> CustomStyling
+                    "java-example" -> Java
+                    "kotlin-as-java-example" -> KotlinAsJava
+                    "kotlin-multiplatform-example" -> KotlinMultiplatform
+                    "library-publishing-example" -> LibraryPublishing
+                    "multimodule-example" -> Multimodule
+                    "versioning-multimodule-example" -> VersioningMultimodule
+                    else -> error("Undeclared example project: $dir")
+                }
+            }
+        }
+    }
+
+
+    @ParameterizedTest
+    @ArgumentsSource(TestCaseProvider::class)
+    fun `test HTML output`(testCase: TestCase) {
+        assumeTrue(testCase.outputsHtml)
+
+        testDokkaOutput(
+            testCase = testCase,
+            format = "html",
+        )
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(TestCaseProvider::class)
+    fun `test Javadoc output`(testCase: TestCase) {
+        assumeTrue(testCase.outputsJavadoc)
+
+        testDokkaOutput(
+            testCase = testCase,
+            format = "javadoc",
+        )
+    }
+
+    private fun testDokkaOutput(
+        testCase: TestCase,
+        format: String,
+    ) {
+        val expectedDataDir = testCase.expectedDataDir.resolve(format)
+        val dokkaOutputDir = testCase.dokkaOutputDir.resolve(format)
+
+        assert(expectedDataDir.exists() && expectedDataDir.isDirectory()) {
+            "Missing expectedDataDir: ${expectedDataDir.toUri()}"
+        }
+
+        testCase.project.runner
+            .addArguments(
+                testCase.dokkaGenerateTask,
+                "--stacktrace",
+            )
+            .build {
+                dokkaOutputDir.shouldBeADirectory()
+
+                withClue({
+                    """
+                    expectedDataDir: ${expectedDataDir.toUri()}
+                    actualOutputDir: ${dokkaOutputDir.toUri()}
+                    """.trimIndent()
+                }) {
+                    withClue("expect file trees are the same") {
+                        val expectedFileTree = expectedDataDir.toTreeString()
+                        val actualFileTree = dokkaOutputDir.toTreeString()
+                        expectedFileTree shouldBe actualFileTree
+                    }
+
+                    withClue("expect directories are the same") {
+                        dokkaOutputDir.shouldHaveSameStructureAs(expectedDataDir, skipEmptyDirs = true)
+                        dokkaOutputDir.shouldHaveSameStructureAndContentAs(expectedDataDir, skipEmptyDirs = true)
```

Comment:
Do we need `shouldHaveSameStructureAs` and `shouldHaveSameStructureAndContentAs` simultaneously?

nb: `shouldHaveSameStructureAs` calls `shouldHaveSameStructureAndContentAs2` overload which doesn't actually compare contents

Reply:
It's not a big difference, but I found the `shouldHaveSameStructureAs` assertion failure message is more useful.